Images are not drawn even after they are completely loaded from a slow server

shallawa opened the pull request:
#### b56577295e3d7564c3f3b74f0bd6cdb7647569e0
<pre>
Images are not drawn even after they are completely loaded from a slow server
<a href="https://bugs.webkit.org/show_bug.cgi?id=249645">https://bugs.webkit.org/show_bug.cgi?id=249645</a>
rdar://103560731

Reviewed by Simon Fraser.

While loading a BitmapImage (i.e. allDataReceived is false), a partially decoded
NativeImage is created and cached in ImageFrame with DecodingStatus::Partial.

After loading the image is completed, ImageSource::cacheMetadataAtIndex() gets
called with the default DecodingStatus which is DecodingStatus::Invalid. This
forces recalculating a new DecodingStatus which will be DecodingStatus::Complete.

So the ImageFrame will have partially decoded NativeImage and DecodingStatus::
Complete. The next draw ImageFrame::hasDecodedNativeImageCompatibleWithOptions()
will return true and a stale NativeImage is drawn, which usually draws nothing
except for JPEG images.

It turns out we were setting ImageFrame::m_decodingStatus incorrectly always when
decoding an image asynchronously on GTK and WPE ports. We get the decodingStatus
of the frame when requesting it to be asynchronously decoded. This is calculated
by calling frameIsCompleteAtIndex(). ScalableImageDecoder::frameIsCompleteAtIndex()
returns false always if we have not decoded the frame yet.

The fix is to let ImageSource::cachePlatformImageAtIndex() be the only place to
set the ImageFrame::m_decodingStatus since it is the only setter for the
ImageFrame::nativeImage and we are sure at this point frameIsCompleteAtIndex()
will return the correct value for all ports.

* LayoutTests/http/tests/images/render-partial-image-load-expected.html: Added.
* LayoutTests/http/tests/images/render-partial-image-load.html: Added.
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::hasDecodedNativeImageCompatibleWithOptions const):
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::cacheMetadataAtIndex):
(WebCore::ImageSource::cachePlatformImageAtIndex):
(WebCore::ImageSource::cachePlatformImageAtIndexAsync):
(WebCore::ImageSource::startAsyncDecodingQueue):
(WebCore::ImageSource::requestFrameAsyncDecodingAtIndex):
(WebCore::ImageSource::frameAtIndexCacheIfNeeded):
* Source/WebCore/platform/graphics/ImageSource.h:

Canonical link: <a href="https://commits.webkit.org/263085@main">https://commits.webkit.org/263085@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34bdc156878a98717e05fafa57e4e3219f695c10

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3408 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3466 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3594 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4835 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3724 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3387 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3541 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3498 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2968 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3450 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3719 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3073 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4655 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1227 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3046 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2956 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3015 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3104 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4403 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3477 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2793 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3031 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3044 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/866 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3045 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3308 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/http/tests/resources/boston.gif:
Do we really need to add a new image resource for this? Can we use one of the existing resources in this directory?

shallawa replied:
The images was removed.

shallawa committed:
Images are not drawn even after they are completely loaded from a slow server
https://bugs.webkit.org/show_bug.cgi?id=249645
rdar://103560731

Reviewed by Simon Fraser.

While loading a BitmapImage (i.e. allDataReceived is false), a partially decoded
NativeImage is created and cached in ImageFrame with DecodingStatus::Partial.

After loading the image is completed, ImageSource::cacheMetadataAtIndex() gets
called with the default DecodingStatus which is DecodingStatus::Invalid. This
forces recalculating a new DecodingStatus which will be DecodingStatus::Complete.

So the ImageFrame will have partially decoded NativeImage and DecodingStatus::
Complete. The next draw ImageFrame::hasDecodedNativeImageCompatibleWithOptions()
will return true and a stale NativeImage is drawn, which usually draws nothing
except for JPEG images.

It turns out we were setting ImageFrame::m_decodingStatus incorrectly always when
decoding an image asynchronously on GTK and WPE ports. We get the decodingStatus
of the frame when requesting it to be asynchronously decoded. This is calculated
by calling frameIsCompleteAtIndex(). ScalableImageDecoder::frameIsCompleteAtIndex()
returns false always if we have not decoded the frame yet.

The fix is to let ImageSource::cachePlatformImageAtIndex() be the only place to
set the ImageFrame::m_decodingStatus since it is the only setter for the
ImageFrame::nativeImage and we are sure at this point frameIsCompleteAtIndex()
will return the correct value for all ports.

* LayoutTests/http/tests/images/render-partial-image-load-expected.html: Added.
* LayoutTests/http/tests/images/render-partial-image-load.html: Added.
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::hasDecodedNativeImageCompatibleWithOptions const):
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::cacheMetadataAtIndex):
(WebCore::ImageSource::cachePlatformImageAtIndex):
(WebCore::ImageSource::cachePlatformImageAtIndexAsync):
(WebCore::ImageSource::startAsyncDecodingQueue):
(WebCore::ImageSource::requestFrameAsyncDecodingAtIndex):
(WebCore::ImageSource::frameAtIndexCacheIfNeeded):
* Source/WebCore/platform/graphics/ImageSource.h:

Canonical link: https://commits.webkit.org/263085@main
